perception_dataset/t4_dataset/data_interpolator.py

```diff
@@ -0,0 +1,497 @@
+        self.logger.info("Finish updating scene")
+
+        # save
+        annotation_root = osp.join(output_path, nusc.version)
+        self._save_json(all_samples, osp.join(annotation_root, "sample.json"))
+        self._save_json(all_sample_data, osp.join(annotation_root, "sample_data.json"))
+        self._save_json(all_sample_anns, osp.join(annotation_root, "sample_annotation.json"))
+        self._save_json(all_instances, osp.join(annotation_root, "instance.json"))
+        self._save_json(all_scenes, osp.join(annotation_root, "scene.json"))
+
+    def interpolate_sample(
+        self, nusc: NuScenes
+    ) -> Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
+        """
+        Interpolate sample with non-key frame sample data.
+
+        Args:
+            nusc (NuScenes): NuScenes instance.
+
+        Returns:
+            Tuple[List[Dict[str, Any]], List[Dict[str, Any]]]:
+                All sample records containing interpolated ones,
+                and all sample data that `is_key_frame` is converted to `True`.
+        """
+        # interpolate sample record based on lidar timestamp
+        scene_token: str = nusc.sample[0]["scene_token"]
+        interpolated_samples_dict: Dict[str, Dict[str, Any]] = {}
+        for sd_record in nusc.sample_data:
+            if sd_record["is_key_frame"] or sd_record["sensor_modality"] != "lidar":
+                continue
+
+            # [filename, pcd, bin]
+            filename: str = osp.basename(sd_record["filename"]).split(".")[0]
+            interpolated_samples_dict[filename] = {
+                "token": token_hex(16),
+                "timestamp": sd_record["timestamp"],
+                "scene_token": scene_token,
+                "prev": None,
+                "next": None,
+            }
+
+        # update prev/next token in sample record
+        all_sorted_samples = sorted(
+            [{key: s[key] for key in self.SAMPLE_KEYS} for s in nusc.sample]
+            + list(interpolated_samples_dict.values()),
+            key=lambda s: s["timestamp"],
+        )
+        prev_token: str = ""
+        for i, sample_record in enumerate(all_sorted_samples[:-1]):
+            next_sample_token = all_sorted_samples[i + 1]["token"]
+            sample_record["prev_token"] = prev_token
+            sample_record["next"] = next_sample_token
+            prev_token = sample_record["token"]
+        all_sorted_samples[-1]["prev"] = prev_token
+        all_sorted_samples[-1]["next"] = ""
+
+        # update sample_token in sample_data
+        all_sample_data = [
+            {name: record[name] for name in self.SAMPLE_DATA_KEYS if name in record.keys()}
+            for record in nusc.sample_data
+        ]
+        for sd_record in all_sample_data:
+            if sd_record["is_key_frame"] or "pcd.bin" == sd_record["fileformat"]:
+                continue
+            filename: str = osp.splitext(osp.basename(sd_record["filename"]))[0]
```

Comment:
[Q]
Is `"pcd.bin" == sd_record["fileformat"]` necessary?
I think this should be removed in order to interpolate LiDAR data.
```suggestion
            if sd_record["is_key_frame"]:
                continue
            filename: str = osp.splitext(osp.basename(sd_record["filename"].replace(".pcd.bin", ".pcd")))[0]
```